platforms.yaml

```diff
@@ -99,6 +118,17 @@ x86_64:
     kernel_arguments:
       - console=tty0
       - console=ttyS0,115200n8
+  openstack:
+    # Graphical console primary, serial console available for logging
+    # https://docs.openstack.org/diskimage-builder/latest/elements/bootloader/README.html
+    # https://issues.redhat.com/browse/OCPBUGS-2926
+    grub_commands:
+      - serial --speed=115200
+      - terminal_input serial console
+      - terminal_output serial console
+    kernel_arguments:
+      - console=ttyS0,115200n8
+      - console=tty0
```

Comment:
this is more or less inline with our documentation 
https://docs.openstack.org/image-guide/centos-image.html#configure-console

a standard centos 9 cloud image has

BOOT_IMAGE=(hd0,msdos1)/boot/vmlinuz-5.14.0-165.el9.x86_64 root=UUID=8db50efa-f8ad-4ee1-a332-07c7abb146c6 ro console=ttyS0,115200n8 no_timer_check net.ifnames=0 crashkernel=1G-4G:192M,4G-64G:256M,64G-:512M

ubuntu 22.04 is a little more minimal 

BOOT_IMAGE=/boot/vmlinuz-5.15.0-48-generic root=LABEL=cloudimg-rootfs ro console=tty1 console=ttyS0

but in general i believe this is more or less correct

in my experience i have not needed to set the grub_commands to enable serial output as in most distros adding the kernel args is enough.

both set 

ubuntu sets

# Set the grub console type
GRUB_TERMINAL=console

and 
centos sets 
GRUB_TERMINAL_OUTPUT="console"

in there cloud images by default
so presumably that is what is being configured by 

  - terminal_input serial console
  - terminal_output serial console

overall i think this should work but its probably more verbose than required

Reply:
Yup, we wanted to preserve graphical console access (`console=tty0`) and allow the GRUB menu to be accessed via serial console.

Reply:
in general dont think you need to do anything special to preserve the graphical console.

in OpenStack we generate XML like this

```
<serial type='tcp'>
      <source mode='bind' host='192.168.1.167' service='10004' tls='no'/>
      <protocol type='raw'/>
      <log file='/var/lib/nova/instances/97b112d3-7702-42b6-97d3-1e582961cd09/console.log' append='off'/>
      <target type='isa-serial' port='0'>
        <model name='isa-serial'/>
      </target>
      <alias name='serial0'/>
    </serial>
    <console type='tcp'>
      <source mode='bind' host='192.168.1.167' service='10004' tls='no'/>
      <protocol type='raw'/>
      <log file='/var/lib/nova/instances/97b112d3-7702-42b6-97d3-1e582961cd09/console.log' append='off'/>
      <target type='serial' port='0'/>
      <alias name='serial0'/>
    </console>
```

we define a serial port then use that as the source for the console

```
  <graphics type='vnc' port='5904' autoport='yes' listen='192.168.1.167'>
      <listen type='address' address='192.168.1.167'/>
    </graphics>
    <audio id='1' type='none'/>
    <video>
      <model type='virtio' heads='1' primary='yes'/>
      <alias name='video0'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x0'/>
    </video>

```

then later we define a graphics device of type vnc or spice that exports that console

i generally have had good experience just setting "console=tty0 console=ttyS0,115200n8"

the kernel args are more or less identical to what i would generally set although i always listed the serial console second.

so as i said the kernel argmumets look correct to me and should work to allow the console to be accessed either via horizon or via the console log command.


the grub_commands I'm less sure about but i don't think they would break anything if included so without testing the final image by botting a VM and installing a graphical ide i don't see any redflags here.

Reply:
Right, it's the `console=tty0` argument that directs the kernel to write to the graphical console.  If you only specify `console=ttyS0,115200n8` it'll only send output to the serial port.  Regarding the order of the `console` arguments, the last argument becomes the primary console and that gets a few extra pieces of output.  Here, I've opted to preserve the full UX of the graphical console, and collect logs on the serial port as best as we can.

For context, this change is part of a larger project to rework the default console settings for each of the supported FCOS/RHCOS platforms.  I think we have a good handle on the general approach; the main thing here is to make sure we've fixed the use case reported in [OCPBUGS-2926](https://issues.redhat.com/browse/OCPBUGS-2926), and it sounds as though we're good on that front too.  Thanks for your feedback! :+1: